src/bernsteinyang.rs

```diff
@@ -0,0 +1,351 @@
+            
+            matrix[1] = [matrix[0][0] * w + matrix[1][0], matrix[0][1] * w + matrix[1][1]];
+            g += w as i128 * f as i128;
+        }
+    
+        (delta, matrix)
+    }
+    
+    fn fg(f: ChunkInt<B,L>, g: ChunkInt<B,L>, matrix: Matrix) -> (ChunkInt<B,L>, ChunkInt<B,L>) {
+        ((matrix[0][0] * &f + matrix[0][1] * &g).shift(), (matrix[1][0] * &f + matrix[1][1] * &g).shift()) 
+    }
+
+    fn de(&self, d: ChunkInt<B,L>, e: ChunkInt<B,L>, matrix: Matrix) -> (ChunkInt<B,L>, ChunkInt<B,L>) {
+        let mask = ChunkInt::<B,L>::MASK as i64;
+        let mut md = matrix[0][0] * d.is_negative() as i64 + matrix[0][1] * e.is_negative() as i64;
+        let mut me = matrix[1][0] * d.is_negative() as i64 + matrix[1][1] * e.is_negative() as i64;
+
+        let cd = matrix[0][0].wrapping_mul(d.lowest() as i64).wrapping_add(matrix[0][1].wrapping_mul(e.lowest() as i64)) & mask; 
+        let ce = matrix[1][0].wrapping_mul(d.lowest() as i64).wrapping_add(matrix[1][1].wrapping_mul(e.lowest() as i64)) & mask;
+
+        md -= (self.inverse.wrapping_mul(cd).wrapping_add(md)) & mask;
+        me -= (self.inverse.wrapping_mul(ce).wrapping_add(me)) & mask;
+        
+        let cd = matrix[0][0] * &d + matrix[0][1] * &e + md * &self.modulus;
+        let ce = matrix[1][0] * &d + matrix[1][1] * &e + me * &self.modulus;
+
+        (cd.shift(), ce.shift())   
+    }
+
+    fn norm(&self, mut value: ChunkInt<B,L>, negate: bool) -> ChunkInt<B,L> {
+        if value.is_negative() {
+            value = value + &self.modulus;
+        }
+
+        if negate {
+            value = -value;
+        }
+
+        if value.is_negative() {
+            value = value + &self.modulus;
+        }
+
+        value
+    }
+
+    /// Returns a big unsigned integer as an array of O-bit chunks, which is equal modulo
+    /// 2 ^ (O * S) to the input big unsigned integer stored as an array of I-bit chunks.
+    /// The ordering of the chunks in these arrays is little-endian
+    const fn convert<const I:usize, const O:usize, const S:usize>(input: &[u64]) -> [u64; S] {
+        const fn min(a: usize, b: usize) -> usize { if a > b { b } else { a } }
+        let (total, mut output, mut bits) = (min(input.len() * I, S * O), [0; S], 0);
+          
+        while bits < total {
+            let (i, o) = (bits % I, bits % O);
+            output[bits / O] |= (input[bits / I] >> i) << o;
+            bits += min(I - i, O - o);
+        }
+        
+        let mask = u64::MAX >> (64 - O);
+        let mut filled = total / O + if total % O > 0 { 1 } else { 0 };
+    
+        while filled > 0 {
+            filled -= 1;
+            output[filled] &= mask;
+        }
+        
+        output
+    }
+
+    /// Returns the multiplicative inverse of the argument modulo 2^B. The implementation is based
+    /// on the Hurchalla's method for computing the multiplicative inverse modulo a power of two
```

Comment:
Missing link to paper

Reply:
Usually I try to avoid overwhelming my code with direct links to papers easy to find, since I try to achieve better readability and aesthetics) However, I can add this one.